src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkRuleClassFunctions.java

```diff
@@ -1049,13 +1027,62 @@ public ExecGroup execGroup(
       return ExecGroup.copyFromDefault();
     }
 
-    ImmutableSet<Label> toolchainTypes = ImmutableSet.copyOf(parseToolchains(toolchains, thread));
+    ImmutableSet<ToolchainTypeRequirement> toolchainTypes = parseToolchainTypes(toolchains, thread);
     ImmutableSet<Label> constraints =
         ImmutableSet.copyOf(parseExecCompatibleWith(execCompatibleWith, thread));
     return ExecGroup.builder()
-        .requiredToolchains(toolchainTypes)
+        .toolchainTypes(toolchainTypes)
         .execCompatibleWith(constraints)
         .copyFrom(null)
         .build();
   }
+
+  private static ImmutableSet<ToolchainTypeRequirement> parseToolchainTypes(
+      Sequence<?> rawToolchains, StarlarkThread thread) throws EvalException {
+    ImmutableSet.Builder<ToolchainTypeRequirement> toolchainTypes = new ImmutableSet.Builder<>();
+    Set<Label> toolchainTypesSeen = new HashSet<>();
+    LabelConverter converter = LabelConverter.forThread(thread);
+
+    for (Object rawToolchain : rawToolchains) {
+      ToolchainTypeRequirement toolchainType = parseToolchainType(converter, rawToolchain);
+      if (toolchainTypesSeen.contains(toolchainType.toolchainType())) {
+        // throw an error for duplicates.
+        throw Starlark.errorf("Duplicate toolchain type %s", toolchainType.toolchainType());
```

Comment:
I'm actually not sure about this, we may need to instead merge TTRs, keeping the most strict. Thoughts?

Reply:
Mandatory is more strict than optional. So if a rule requests `(@bazel_tools//tools/cpp:toolchain_type, Optional)` and `(@bazel_tools//tools/cpp:toolchain_type, Mandatory)`, we'd collapse that to just the mandatory version.

Reply:
I see. Toolchain types can't be inherited, can they? Is there a Starlark mechanism for rule definitions to inherit from other rule definitions?

With inheritance I could see an argument for allowing further locking down --> turning an optional into a required but not vice versa. 

Without inheritance, is there a good reason to actually have that repetition? I'd be inclined to be strict on that. We can always of course loosen if people run into problem and actually need the more complicated logic.

Reply:
There's no way to do inheritance in Starlark like there is in native. There may be cases where rule authors provide helpers to create the rule configurations and they combine oddly, however.

The more I think, the more I like deduping instead of failing.

Reply:
Why?

Reply:
In code, the toolchain type requirements act like a set (and are actually implemented with ImmutableSet). I think this is cleaner, and easier for the user (rule author) to understand.

Reply:
I'm missing the scenario in which the user would set the same toolchain for both types. Why is that more than just obvious repetition? Bazel, for example, reports an error on

```
cc_binary(
    name = "appl",
    srcs = ["main.c", "main.c"],
)
```

But I'm hearing this isn't the right comparison?

Reply:
A lot of rules are written by composing helpers, and those helpers are not always completely independent. We currently have many places where rules are creating the `attrs` argument by creating a dict and using that to handle merging name conflicts, so this is something that happens.

I've updated the code to merge toolchain type requirements. In this case, mandatory requirements are considered stricter than optional, since they have tighter constraints, and so that if both are present the requirement is considered to be mandatory.

Reply:
Thanks for all the elaboration. I don't want to block progress on this and your explanation makes sense. Resolving this thread and I'll do another review over the current state of the code.